File: exam-system-java/server/src/main/java/com/yyjy/exam/question/common/QuestionExcelListener.java
package com.yyjy.exam.question.common;

import com.alibaba.excel.context.AnalysisContext;
import com.alibaba.excel.event.AnalysisEventListener;
import com.yyjy.exam.entity.question.dto.QuestionImportView;
import com.yyjy.exam.entity.question.entity.QuestionDifficulty;
import com.yyjy.exam.entity.question.entity.QuestionType;
import com.yyjy.exam.question.bo.QuestionExcelTemplateBo;

import java.util.ArrayList;
import java.util.List;

public class QuestionExcelListener extends AnalysisEventListener<QuestionExcelTemplateBo> {
	
	private final List<QuestionImportView> datas = new ArrayList<>();
	private final List<String> errors = new ArrayList<>();
	
	@Override
	public void invoke(QuestionExcelTemplateBo data, AnalysisContext context) {
		List<String> validationErrors = validateData(datas.size() + 1, data.getContent(), data.getType(), data.getAnswer(), data.getMultiple());
		if (!validationErrors.isEmpty()) {
			errors.addAll(validationErrors);
			return;
		}
		
		QuestionImportView question = new QuestionImportView();
		question.setTitle(data.getContent());
		question.setType(QuestionType.valueOf(data.getType()));
		question.setCategoryId(Long.parseLong(data.getCategoryId()));
		question.setDifficulty(QuestionDifficulty.valueOf(data.getDifficulty()));
		question.setScore(parseIntOrDefault(data.getScore()));
		question.setAnalysis(data.getAnalysis());
		question.setAnswer(data.getAnswer());
		
		if (QuestionType.SINGLE_CHOICE.toString().equals(data.getType()) || QuestionType.MULTIPLE_CHOICE.toString().equals(data.getType())) {
			String[] choiceContents = {data.getChoiceA(), data.getChoiceB(), data.getChoiceC(), data.getChoiceD()};
			String answer = data.getAnswer();
			List<QuestionImportView.TargetOf_choices> choicesList = new ArrayList<>();
			for (int i = 0; i < choiceContents.length; i++) {
				if (choiceContents[i] != null) {
					QuestionImportView.TargetOf_choices c = new QuestionImportView.TargetOf_choices();
					c.setContent(choiceContents[i]);
					c.setCorrect(answer != null && answer.contains(choiceContents[i]));
					c.setSort(i);
					choicesList.add(c);
				}
			}
			question.setChoices(choicesList);
		}
		
		datas.add(question);
	}
	
	public List<QuestionImportView> getDatas() {
		return datas;
	}
	
	public List<String> getErrors() {
		return errors;
	}
	
	@Override
	public void doAfterAllAnalysed(AnalysisContext context) {
	}
	
	private int parseIntOrDefault(String value) {
		if (value == null) return 1;
		try {
			return Integer.parseInt(value);
		} catch (NumberFormatException e) {
			return 1;
		}
	}
	
	private List<String> validateData(int row, String content, String type, String answer, String multiple) {
		List<String> errors = new ArrayList<>();
		String prefix = row + " 行数据";
		
		if (content == null || content.isBlank()) errors.add(prefix + ": 缺失题目");
		if (type == null || type.isBlank()) errors.add(prefix + ": 缺失题目类型");
		if (answer == null || answer.isBlank()) errors.add(prefix + ": 缺失答案");
		
		if (!"是".equals(multiple) && !"否".equals(multiple)) {
			errors.add(prefix + ": 选择题答案必须为是或否");
		}
		
		if (QuestionType.JUDGE.toString().equals(type)) {
			if (!"true".equals(answer) && !"false".equals(answer)) {
				errors.add(prefix + ": 判断题答案必须为true或false");
			}
		}
		
		if (QuestionType.SINGLE_CHOICE.toString().equals(type) && answer != null) {
			if (answer.length() != 1 || answer.charAt(0) < 'A' || answer.charAt(0) > 'D') {
				errors.add(prefix + ": 选择题答案只能为A-D的字母");
			}
		}
		
		if (QuestionType.MULTIPLE_CHOICE.toString().equals(type) && answer != null) {
			if (!answer.matches("^[A-D](,[A-D])+$")) {
				errors.add(prefix + ": 选择题答案只能为A-D的字母，多个答案用逗号隔开");
			}
		}
		
		return errors;
	}
}

File: exam-system-server/server/src/main/kotlin/com/yyjy/common/QuestionExcelListener.kt
package com.yyjy.common

import com.alibaba.excel.context.AnalysisContext
import com.alibaba.excel.event.AnalysisEventListener
import com.yyjy.constants.QuestionConstant
import com.yyjy.models.bo.QuestionExcelTemplateBo
import com.yyjy.models.entity.dto.QuestionImportView

/**
 * @author Nyxcirea
 * @date 2026/4/11
 * @description: TODO
 */
class QuestionExcelListener : AnalysisEventListener<QuestionExcelTemplateBo>() {
    //可以通过实例获取该值
    private val datas = mutableListOf<QuestionImportView>()
    private val errors = mutableListOf<String>()  // 收集所有错误
    /**
     * 监听到每一条数据都会执行此方法
     */
    override fun invoke(p0: QuestionExcelTemplateBo, analysisContext: AnalysisContext?) {

        // 校验数据，不抛异常，只收集错误
        val validationErrors = ValidateExcel.validateData(datas.size + 1, p0.content, p0.type, p0.answer, p0.multiple)

        if (validationErrors.isNotEmpty()) {
            // 有错误，记录但不处理这条数据
            errors.addAll(validationErrors)
            return  // 跳过这条数据
        }

        //数据存储到list，供批量处理，或后续自己业务逻辑处理。
        val question = QuestionImportView(
            title = p0.content!!,
            type = p0.type!!,
            multi = p0.multiple == "是",
            categoryId = p0.categoryId!!.toLong(),
            difficulty = p0.difficulty!!,
            score = p0.score?.toIntOrNull(),
            analysis = p0.analysis,
            answer = p0.answer,
        )
        val choices = mutableListOf<QuestionImportView.TargetOf_choices>()
        // 处理选择题的选项，判断题和简答题不需选项
        if (p0.type == QuestionConstant.TYPE.CHOICE) {
            val answer = p0.answer
            // 处理选项
            for ((index, item) in listOf(p0.choiceA, p0.choiceB, p0.choiceC, p0.choiceD).withIndex()) {
                choices.add(QuestionImportView.TargetOf_choices(
                    content = item!!,
                    sort = index,
                    correct = answer!!.contains(item)
                ))
            }
            question.choices = choices
        }
        datas.add(question)
    }

    fun getDatas(): List<QuestionImportView> {
        return datas
    }
    fun getErrors(): List<String> {
        return errors
    }
    /**
     * 解析完成之后执行此方法
     */
    override fun doAfterAllAnalysed(analysisContext: AnalysisContext?) {
        // 解析完成后，如果有错误，统一抛出
//        if (errors.isNotEmpty()) {
//            val errorMessage = errors.joinToString("\n")
//            throw BusinessException("数据校验失败：\n$errorMessage")
//        }
    }
}
object ValidateExcel {
    fun validateData(dataLen: Int, content: String?, type: String?, answer: String?, multiple: String?): List<String> {
        val errors = mutableListOf<String>()
        val prefix = "$dataLen 行数据"

        // 基础字段校验
        if (content.isNullOrBlank()) errors.add("$prefix: 缺失题目")
        if (type.isNullOrBlank()) errors.add("$prefix: 缺失题目类型")
        if (answer.isNullOrBlank()) errors.add("$prefix: 缺失答案")

        // multiple 字段校验
        if (multiple != "是" && multiple != "否") {
            errors.add("$prefix: 选择题答案必须为是或否")
        }

        // 根据题型校验答案
        when (type) {
            QuestionConstant.TYPE.JUDGE -> validateJudgeAnswer(answer, prefix, errors)
            QuestionConstant.TYPE.CHOICE -> validateChoiceAnswer(answer, multiple, prefix, errors)
        }

        return errors
    }

    private fun validateJudgeAnswer(answer: String?, prefix: String, errors: MutableList<String>) {
        if (answer != "true" && answer != "false") {
            errors.add("$prefix: 判断题答案必须为true或false")
        }
    }

    private fun validateChoiceAnswer(answer: String?, multiple: String?, prefix: String, errors: MutableList<String>) {
        if (answer.isNullOrBlank()) return

        when (multiple) {
            "是" -> validateMultipleChoice(answer, prefix, errors)
            "否" -> validateSingleChoice(answer, prefix, errors)
        }
    }

    private fun validateMultipleChoice(answer: String, prefix: String, errors: MutableList<String>) {
        val regex = Regex("^[A-D](,[A-D])+$")
        if (!regex.matches(answer)) {
            errors.add("$prefix: 选择题答案只能为A-D的字母，多个答案用逗号隔开")
        }
    }

    private fun validateSingleChoice(answer: String, prefix: String, errors: MutableList<String>) {
        if (answer.length != 1 || answer[0] !in 'A'..'D') {
            errors.add("$prefix: 选择题答案只能为A-D的字母")
        }
    }
}